File: src/render.c
#include "render.h"

#include <stdio.h>

void print_game_grid(const CardinalChainsGame *game, int current_chain,
                     int active_chains[]) {
    const char *chain_colors[] = {"\033[31m", "\033[32m", "\033[33m",
                                  "\033[34m", "\033[35m", "\033[36m"};
    const char *reset_color = "\033[0m";

    for (int i = 0; i < game->level.count; i++) {
        for (int j = 0; j < game->level.value_lengths[i]; j++) {
            bool is_part_of_chain = false;
            int chain_index;

            for (int k = 0; k < game->chain_count; k++) {
                for (int l = 0; l < game->chain_lengths[k]; l++) {
                    if (game->chains[k][l].x == i && game->chains[k][l].y == j) {
                        is_part_of_chain = true;
                        chain_index = active_chains[k];
                        break;
                    }
                }
                if (is_part_of_chain)
                    break;
            }

            if (is_part_of_chain) {
                printf("%s", chain_colors[chain_index % (sizeof(chain_colors) /
                                                         sizeof(chain_colors[0]))]);
                int value = game->level.values[i][j];
                printf("%c ", value == -1 ? 'x' : '0' + value);
                printf("%s", reset_color);
            } else if (game->level.values[i][j] != 0) {
                printf("%d ", game->level.values[i][j]);
            } else {
                printf("  ");
            }
        }
        printf("\n");
    }

    printf("Current position: row %d, column %d\n",
           game->chains[current_chain][game->chain_lengths[current_chain] - 1].x +
           1,
           game->chains[current_chain][game->chain_lengths[current_chain] - 1].y +
           1);
    printf("Current chain color: %sChain %d%s\n",
           chain_colors[active_chains[current_chain] %
                        (sizeof(chain_colors) / sizeof(chain_colors[0]))],
           active_chains[current_chain] + 1, reset_color);
}

File: go/render/render.go
package render

import (
	"fmt"
	"strings"

	"cardinal-chains/game"
)

// chainColors maps a chain color index to a tview color tag. These mirror
// the 6 ANSI colors used in the C render.c implementation:
//
//	red, green, yellow, blue, magenta, cyan
var chainColors = []string{
	"[red]",
	"[green]",
	"[yellow]",
	"[blue]",
	"[magenta]",
	"[cyan]",
}

// colorReset is the tview tag to reset color formatting.
const colorReset = "[-:]"

// chainColor returns the tview color tag for the given color index, cycling
// through the available colors.
func chainColor(idx int) string {
	return chainColors[idx%len(chainColors)]
}

// colorizedName returns a human-readable color name for a color index.
func colorizedName(idx int) string {
	names := []string{"Red", "Green", "Yellow", "Blue", "Magenta", "Cyan"}
	return names[idx%len(names)]
}

// chainIndexAt returns the color index (from activeChains) of the chain
// occupying cell (row, col), or (-1, false) if no chain is there.
func chainIndexAt(g *game.Game, row, col int, activeChains []int) (int, bool) {
	for k, ch := range g.Chains {
		for _, pos := range ch {
			if pos.X == row && pos.Y == col {
				if k < len(activeChains) {
					return activeChains[k], true
				}
				return k, true
			}
		}
	}
	return -1, false
}

// GridString renders the entire game grid as a string suitable for a tview
// TextView. Chains are colored using tview color tags.
func GridString(g *game.Game, currentChain int, activeChains []int) string {
	var sb strings.Builder

	for i := range g.Level.Values {
		for j := range g.Level.Values[i] {
			value := g.Level.Values[i][j]

			colorIdx, isChain := chainIndexAt(g, i, j, activeChains)
			if isChain {
				sb.WriteString(chainColor(colorIdx))
				if value == -1 {
					sb.WriteString("x ")
				} else {
					sb.WriteString(fmt.Sprintf("%d ", value))
				}
				sb.WriteString(colorReset)
			} else if value != 0 {
				sb.WriteString(fmt.Sprintf("%d ", value))
			} else {
				sb.WriteString("  ")
			}
		}
		sb.WriteString("\n")
	}

	return sb.String()
}

// StatusString renders the status line showing the current position and
// current chain color, mirroring the C render.c output.
func StatusString(g *game.Game, currentChain int, activeChains []int) string {
	var sb strings.Builder

	if len(g.Chains) == 0 || currentChain < 0 || currentChain >= len(g.Chains) {
		return ""
	}

	chain := g.Chains[currentChain]
	lastPos := chain[len(chain)-1]

	sb.WriteString(fmt.Sprintf("Current position: row %d, column %d\n",
		lastPos.X+1, lastPos.Y+1))

	colorIdx := currentChain
	if currentChain < len(activeChains) {
		colorIdx = activeChains[currentChain]
	}

	sb.WriteString(fmt.Sprintf("Current chain color: %sChain %d%s (%s)\n",
		chainColor(colorIdx), colorIdx+1, colorReset, colorizedName(colorIdx)))

	return sb.String()
}

// HelpString returns the help text listing all available commands.
func HelpString() string {
	return strings.Join([]string{
		"Commands:",
		"  N/S/E/W - Move the current chain",
		"  B       - Cancel (undo) the previous move",
		"  R       - Erase the current chain back to its anchor",
		"  X       - Restart the whole level",
		"  C       - Cycle selection to the next chain",
		"  Q       - Quit the game",
	}, "\n")
}

// FullRender combines the grid, status, and help into a single string.
func FullRender(g *game.Game, currentChain int, activeChains []int) string {
	var sb strings.Builder
	sb.WriteString(GridString(g, currentChain, activeChains))
	sb.WriteString("\n")
	sb.WriteString(StatusString(g, currentChain, activeChains))
	sb.WriteString("\n")
	sb.WriteString(HelpString())
	return sb.String()
}
